File: resources/android/OtaFunctions.kt
package com.nativephp.plugins.mobile_ota

import android.app.AlertDialog
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.content.SharedPreferences
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeError
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL

object OtaFunctions {
    private const val PREFS = "nativephp_ota"
    private const val KEY_VERSION = "version"

    private fun prefs(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    // Plugin-owned backup of the last pending zip. Must NOT live under
    // app_storage/updates — core applies any *.zip there on the next boot.
    private fun otaDir(context: Context): File =
        File(context.filesDir, "ota").apply { mkdirs() }

    private fun previousZip(context: Context) = File(otaDir(context), "previous.zip")

    // Core pending location: {appStorageDir}/updates/pending.zip
    // Matches LaravelEnvironment.appStorageDir = context.getDir("storage", MODE_PRIVATE)
    private fun appStorageDir(context: Context): File =
        context.getDir("storage", Context.MODE_PRIVATE)

    private fun updatesDir(context: Context): File =
        File(appStorageDir(context), "updates").apply { mkdirs() }

    private fun pendingZip(context: Context) = File(updatesDir(context), "pending.zip")

    // Written after the zip, so its absence marks an interrupted download.
    private fun pendingManifest(context: Context) = File(updatesDir(context), "pending.json")

    private fun installedVersion(context: Context): String {
        val p = prefs(context)
        return try {
            p.getString(KEY_VERSION, null)?.takeIf { it.isNotEmpty() }
                ?: p.getInt(KEY_VERSION, 0).toString()
        } catch (_: ClassCastException) {
            try {
                p.getInt(KEY_VERSION, 0).toString()
            } catch (_: ClassCastException) {
                "0"
            }
        }
    }

    private fun storeVersion(context: Context, version: String) {
        prefs(context).edit().putString(KEY_VERSION, version).apply()
    }

    private fun versionParam(parameters: Map<String, Any>, context: Context): String {
        return when (val v = parameters["version"]) {
            is String -> v.ifEmpty { installedVersion(context) }
            is Number -> v.toString()
            else -> installedVersion(context)
        }
    }

    // Copy the current pending zip aside before it is overwritten so rollback
    // can re-drop it as pending for the next core boot.
    private fun backupPendingIfPresent(context: Context) {
        val pending = pendingZip(context)
        if (!pending.exists()) return
        val previous = previousZip(context)
        if (previous.exists()) previous.delete()
        pending.copyTo(previous, overwrite = true)
    }

    class Check(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            return runCatching { checkForUpdate(context, parameters) }
                .getOrElse { unavailable(versionParam(parameters, context), it.message ?: "check failed") }
        }
    }

    class Download(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val url = parameters["url"] as? String
                ?: return BridgeResponse.error(BridgeError.InvalidParameters("url is required"))
            return runCatching {
                backupPendingIfPresent(context)
                val pending = pendingZip(context)
                val bytes = URL(url).openStream().use { it.readBytes() }

                // A payload that does not match what the server described is not
                // the release we were offered, so it never reaches the location
                // core extracts from.
                val expectedSha = parameters["sha256"] as? String
                if (!expectedSha.isNullOrBlank()) {
                    val actual = java.security.MessageDigest.getInstance("SHA-256")
                        .digest(bytes)
                        .joinToString("") { "%02x".format(it) }
                    if (!actual.equals(expectedSha, ignoreCase = true)) {
                        return BridgeResponse.error(BridgeError.ExecutionFailed("checksum mismatch"))
                    }
                }
                val expectedSize = (parameters["size"] as? Number)?.toInt() ?: 0
                if (expectedSize > 0 && bytes.size != expectedSize) {
                    return BridgeResponse.error(BridgeError.ExecutionFailed("size mismatch: expected $expectedSize, got ${bytes.size}"))
                }

                pending.writeBytes(bytes)

                // What the server said about these bytes, written after them:
                // core treats its absence as an interrupted download, and moves
                // it into the app as ota.json once the payload is applied. The
                // signed download URL is deliberately not persisted.
                (parameters["release"] as? String)?.takeIf { it.isNotBlank() }?.let { release ->
                    val manifest = JSONObject().put("release_uuid", release)
                    for (key in listOf("sha256", "size", "commit", "published_at", "arc", "shell_fingerprint")) {
                        parameters[key]?.let { manifest.put(key, it) }
                    }
                    pendingManifest(context).writeText(manifest.toString(2))
                }

                when (val v = parameters["version"]) {
                    is String -> if (v.isNotEmpty()) storeVersion(context, v)
                    is Number -> storeVersion(context, v.toString())
                }
                BridgeResponse.success(mapOf(
                    "success" to true,
                    "path" to pending.absolutePath,
                    "queued" to true,
                    "applyOnNextBoot" to true
                ))
            }.getOrElse { BridgeResponse.error(BridgeError.ExecutionFailed(it.message ?: "download failed")) }
        }
    }

    class Apply(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val pending = pendingZip(context)
            if (!pending.exists()) {
                return BridgeResponse.error(BridgeError.ExecutionFailed("no pending payload"))
            }
            val version = versionParam(parameters, context)
            storeVersion(context, version)
            // Core extracts {appStorageDir}/updates/*.zip on the next boot.
            // Do not unzip into the running Laravel tree from the plugin.
            return BridgeResponse.success(mapOf(
                "success" to true,
                "version" to version,
                "current_version" to version,
                "queued" to true,
                "applyOnNextBoot" to true,
                "restartRequired" to true
            ))
        }
    }

    class Rollback(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val pending = pendingZip(context)
            val previous = previousZip(context)
            if (!previous.exists()) {
                return BridgeResponse.error(BridgeError.ExecutionFailed("no previous payload"))
            }
            val tmp = File(otaDir(context), "swap.zip")
            if (pending.exists()) pending.copyTo(tmp, overwrite = true)
            previous.copyTo(pending, overwrite = true)
            if (tmp.exists()) {
                tmp.copyTo(previous, overwrite = true)
                tmp.delete()
            }
            val version = installedVersion(context)
            return BridgeResponse.success(mapOf(
                "success" to true,
                "version" to version,
                "current_version" to version,
                "queued" to true,
                "applyOnNextBoot" to true,
                "restartRequired" to true
            ))
        }
    }

    class GetStatus(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val version = installedVersion(context)
            val pending = pendingZip(context).exists()
            return BridgeResponse.success(mapOf(
                "version" to version,
                "current_version" to version,
                "hasPrevious" to previousZip(context).exists(),
                "pending" to pending,
                "queued" to pending,
                "applyOnNextBoot" to pending
            ))
        }
    }

    class Prompt(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val check = checkForUpdate(activity, parameters)
            val data = check["data"] as? Map<*, *>
            val available = data?.get("available") == true
            if (!available) {
                return BridgeResponse.success(mapOf("available" to false, "accepted" to false))
            }
            var accepted = false
            activity.runOnUiThread {
                AlertDialog.Builder(activity)
                    .setTitle("Update available")
                    .setMessage("Download and apply this OTA update?")
                    .setPositiveButton("Update") { _, _ -> accepted = true }
                    .setNegativeButton("Later", null)
                    .show()
            }
            val currentVersion = (data?.get("current_version") as? String)
                ?: (data?.get("version") as? String)
                ?: ""
            return BridgeResponse.success(mapOf(
                "available" to true,
                "accepted" to accepted,
                "version" to currentVersion,
                "current_version" to currentVersion,
                "download_url" to (data?.get("download_url") ?: ""),
                "url" to (data?.get("url") ?: data?.get("download_url") ?: "")
            ))
        }
    }

    private fun parseUpToDate(value: Any?): Boolean? {
        return when (value) {
            is Boolean -> value
            is Number -> value.toInt() != 0
            is String -> when (value.lowercase()) {
                "1", "true", "yes" -> true
                "0", "false", "no" -> false
                else -> null
            }
            else -> null
        }
    }

    private fun unavailable(version: String, reason: String): Map<String, Any> {
        return BridgeResponse.success(mapOf(
            "available" to false,
            "upToDate" to true,
            "current_version" to version,
            "download_url" to "",
            "version" to version,
            "url" to "",
            "reason" to reason
        ))
    }


    private fun isOnline(context: Context): Boolean {
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            ?: return false
        val network = cm.activeNetwork ?: return false
        val caps = cm.getNetworkCapabilities(network) ?: return false
        return caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
    }

    private fun checkForUpdate(context: Context, parameters: Map<String, Any>): Map<String, Any> {
        val endpoint = parameters["endpoint"] as? String
        val project = parameters["project"] as? String
        val arc = parameters["arc"] as? String
        val fingerprint = parameters["fingerprint"] as? String
        val algorithm = parameters["algorithm"] as? String ?: "1"
        val release = parameters["release"] as? String
        val version = versionParam(parameters, context)

        if (endpoint.isNullOrBlank() || project.isNullOrBlank() || arc.isNullOrBlank() || fingerprint.isNullOrBlank()) {
            return unavailable(version, "missing endpoint, project, arc or fingerprint")
        }
        if (!isOnline(context)) {
            return unavailable(version, "offline")
        }

        // The shell says who it is — app, arc, the fingerprint it was built
        // against — and which release it already holds. The answer is whatever
        // that lane points at.
        val trimmed = endpoint.trimEnd('/')
        val held = if (release.isNullOrBlank()) "" else "/$release"
        val url = "$trimmed/api/v1/apps/$project/$arc/$fingerprint$held?algorithm=$algorithm"
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.setRequestProperty("Accept", "application/json")
        (parameters["token"] as? String)?.takeIf { it.isNotBlank() }?.let {
            connection.setRequestProperty("Authorization", "Bearer $it")
        }
        connection.connectTimeout = 15000
        connection.readTimeout = 15000
        try {
            val code = connection.responseCode
            val stream = if (code in 200..299) connection.inputStream else connection.errorStream
            val text = stream?.bufferedReader()?.use { it.readText() } ?: ""
            if (code !in 200..299) {
                return unavailable(version, "http $code")
            }
            if (text.isBlank()) {
                return unavailable(version, "json missing")
            }
            val body = try {
                JSONObject(text)
            } catch (_: Exception) {
                return unavailable(version, "json missing")
            }
            val upToDate = parseUpToDate(body.opt("up_to_date"))
                ?: return unavailable(version, "unexpected response shape")
            val offered = body.optJSONObject("release") ?: JSONObject()
            val releaseUuid = offered.optString("uuid", "")
            val downloadUrl = if (body.has("download_url") && !body.isNull("download_url")) {
                body.optString("download_url", "")
            } else {
                ""
            }
            val available = !upToDate && downloadUrl.isNotBlank() && releaseUuid.isNotBlank()

            return BridgeResponse.success(mapOf(
                "available" to available,
                "upToDate" to upToDate,
                "requested" to url,
                "status" to code,
                "release" to releaseUuid,
                "sha256" to offered.optString("sha256", ""),
                "size" to offered.optInt("size", 0),
                "commit" to offered.optString("commit", ""),
                "published_at" to offered.optString("published_at", ""),
                "download_url" to downloadUrl,
                "url" to downloadUrl,
                "version" to releaseUuid,
                "current_version" to releaseUuid
            ))
        } catch (e: Exception) {
            return unavailable(version, e.message ?: "check failed")
        } finally {
            connection.disconnect()
        }
    }
}

File: resources/ios/OtaFunctions.swift
import Foundation
import Network
import UIKit

enum OtaFunctions {
    private static let versionKey = "nativephp.ota.version"

    /// Plugin-owned backup of the last pending zip. Must NOT live under
    /// Documents/updates — core applies any *.zip there on the next boot.
    private static var otaDirectory: URL {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dir = appSupport.appendingPathComponent("ota", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    private static var previousZip: URL { otaDirectory.appendingPathComponent("previous.zip") }

    /// Core pending location: Documents/updates/pending.zip
    private static var updatesDirectory: URL {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        let dir = docs.appendingPathComponent("updates", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    private static var pendingZip: URL { updatesDirectory.appendingPathComponent("pending.zip") }

    private static var installedVersion: String {
        get {
            if let s = UserDefaults.standard.string(forKey: versionKey), !s.isEmpty {
                return s
            }
            return String(UserDefaults.standard.integer(forKey: versionKey))
        }
        set { UserDefaults.standard.set(newValue, forKey: versionKey) }
    }

    /// Copy the current pending zip aside before it is overwritten so rollback
    /// can re-drop it as pending for the next core boot.
    private static func backupPendingIfPresent() throws {
        let fm = FileManager.default
        guard fm.fileExists(atPath: pendingZip.path) else { return }
        if fm.fileExists(atPath: previousZip.path) {
            try fm.removeItem(at: previousZip)
        }
        try fm.copyItem(at: pendingZip, to: previousZip)
    }

    class Check: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            return try checkForUpdate(parameters: parameters)
        }
    }

    class Download: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let urlString = parameters["url"] as? String, let url = URL(string: urlString) else {
                return [
                    "success": false,
                    "error": "url is required"
                ]
            }
            do {
                try OtaFunctions.backupPendingIfPresent()
                var request = URLRequest(url: url)
                request.httpMethod = "GET"
                request.timeoutInterval = 120
                let semaphore = DispatchSemaphore(value: 0)
                var body: Data?
                var statusCode = 0
                var requestError: Error?
                URLSession.shared.dataTask(with: request) { data, response, error in
                    body = data
                    statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
                    requestError = error
                    semaphore.signal()
                }.resume()
                _ = semaphore.wait(timeout: .now() + 120)
                if let requestError {
                    return [
                        "success": false,
                        "error": requestError.localizedDescription
                    ]
                }
                guard statusCode == 200, let body, !body.isEmpty else {
                    let snippet = body.flatMap { String(data: $0.prefix(180), encoding: .utf8) } ?? ""
                    return [
                        "success": false,
                        "error": "download failed (HTTP \(statusCode)) \(snippet)"
                    ]
                }
                try body.write(to: OtaFunctions.pendingZip, options: .atomic)
                if let version = OtaFunctions.versionString(from: parameters) {
                    OtaFunctions.installedVersion = version
                }
                return BridgeResponse.success(data: [
                    "success": true,
                    "path": OtaFunctions.pendingZip.path,
                    "queued": true,
                    "applyOnNextBoot": true
                ])
            } catch {
                return [
                    "success": false,
                    "error": error.localizedDescription
                ]
            }
        }
    }

    class Apply: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let fm = FileManager.default
            guard fm.fileExists(atPath: OtaFunctions.pendingZip.path) else {
                return BridgeResponse.error(code: "ota", message: "no pending payload")
            }
            let version = OtaFunctions.versionString(from: parameters) ?? OtaFunctions.installedVersion
            OtaFunctions.installedVersion = version
            // Core extracts Documents/updates/*.zip on the next boot.
            // Do not unzip into the running Laravel tree from the plugin.
            return BridgeResponse.success(data: [
                "success": true,
                "version": version,
                "current_version": version,
                "queued": true,
                "applyOnNextBoot": true,
                "restartRequired": true
            ])
        }
    }

    class Rollback: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let fm = FileManager.default
            guard fm.fileExists(atPath: OtaFunctions.previousZip.path) else {
                return BridgeResponse.error(code: "ota", message: "no previous payload")
            }
            let swap = OtaFunctions.otaDirectory.appendingPathComponent("swap.zip")
            if fm.fileExists(atPath: OtaFunctions.pendingZip.path) {
                if fm.fileExists(atPath: swap.path) {
                    try fm.removeItem(at: swap)
                }
                try fm.copyItem(at: OtaFunctions.pendingZip, to: swap)
            }
            if fm.fileExists(atPath: OtaFunctions.pendingZip.path) {
                try fm.removeItem(at: OtaFunctions.pendingZip)
            }
            try fm.copyItem(at: OtaFunctions.previousZip, to: OtaFunctions.pendingZip)
            if fm.fileExists(atPath: swap.path) {
                if fm.fileExists(atPath: OtaFunctions.previousZip.path) {
                    try fm.removeItem(at: OtaFunctions.previousZip)
                }
                try fm.copyItem(at: swap, to: OtaFunctions.previousZip)
                try fm.removeItem(at: swap)
            }
            return BridgeResponse.success(data: [
                "success": true,
                "version": OtaFunctions.installedVersion,
                "current_version": OtaFunctions.installedVersion,
                "queued": true,
                "applyOnNextBoot": true,
                "restartRequired": true
            ])
        }
    }

    class GetStatus: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let fm = FileManager.default
            let pending = fm.fileExists(atPath: OtaFunctions.pendingZip.path)
            return BridgeResponse.success(data: [
                "version": OtaFunctions.installedVersion,
                "current_version": OtaFunctions.installedVersion,
                "hasPrevious": fm.fileExists(atPath: OtaFunctions.previousZip.path),
                "pending": pending,
                "queued": pending,
                "applyOnNextBoot": pending
            ])
        }
    }

    class Prompt: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let check = try checkForUpdate(parameters: parameters)
            let data = check["data"] as? [String: Any]
            let available = data?["available"] as? Bool ?? false
            if !available {
                return BridgeResponse.success(data: ["available": false, "accepted": false])
            }
            let currentVersion = data?["current_version"] as? String
                ?? data?["version"] as? String
                ?? ""
            return BridgeResponse.success(data: [
                "available": true,
                "accepted": false,
                "version": currentVersion,
                "current_version": currentVersion,
                "download_url": data?["download_url"] ?? "",
                "url": data?["url"] ?? data?["download_url"] ?? "",
                "needsUi": true
            ])
        }
    }

    fileprivate static func versionString(from parameters: [String: Any]) -> String? {
        if let s = parameters["version"] as? String, !s.isEmpty {
            return s
        }
        if let i = parameters["version"] as? Int {
            return String(i)
        }
        if let n = parameters["version"] as? NSNumber {
            return n.stringValue
        }
        return nil
    }

    fileprivate static func parseUpToDate(_ value: Any?) -> Bool? {
        switch value {
        case let b as Bool:
            return b
        case let n as NSNumber:
            return n.boolValue
        case let i as Int:
            return i != 0
        case let s as String:
            let lowered = s.lowercased()
            if ["1", "true", "yes"].contains(lowered) { return true }
            if ["0", "false", "no"].contains(lowered) { return false }
            return nil
        default:
            return nil
        }
    }

    fileprivate static func unavailable(version: String, reason: String) -> [String: Any] {
        return BridgeResponse.success(data: [
            "available": false,
            "upToDate": true,
            "current_version": version,
            "download_url": "",
            "version": version,
            "url": "",
            "reason": reason
        ])
    }


    fileprivate static func isOnline() -> Bool {
        let monitor = NWPathMonitor()
        let queue = DispatchQueue(label: "nativephp.ota.network")
        let semaphore = DispatchSemaphore(value: 0)
        var path: NWPath?
        monitor.pathUpdateHandler = { p in
            path = p
            monitor.cancel()
            semaphore.signal()
        }
        monitor.start(queue: queue)
        _ = semaphore.wait(timeout: .now() + 1.5)
        monitor.cancel()
        return path?.status == .satisfied
    }

    fileprivate static func checkForUpdate(parameters: [String: Any]) throws -> [String: Any] {
        let endpoint = parameters["endpoint"] as? String ?? ""
        let project = parameters["project"] as? String ?? ""
        let version = versionString(from: parameters) ?? installedVersion
        if endpoint.isEmpty || project.isEmpty {
            return unavailable(version: version, reason: "missing endpoint or project")
        }
        if !isOnline() {
            return unavailable(version: version, reason: "offline")
        }
        let trimmed = endpoint.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard let url = URL(string: "\(trimmed)/api/apps/\(project)/ota?version=\(version)") else {
            return unavailable(version: version, reason: "invalid endpoint")
        }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 15
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token = parameters["token"] as? String, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        let semaphore = DispatchSemaphore(value: 0)
        var body: [String: Any] = [:]
        var requestError: Error?
        var statusCode = 0
        var parsedJson = false
        URLSession.shared.dataTask(with: request) { data, response, error in
            requestError = error
            if let http = response as? HTTPURLResponse {
                statusCode = http.statusCode
            }
            if let data, let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                body = json
                parsedJson = true
            }
            semaphore.signal()
        }.resume()
        semaphore.wait()
        if let requestError {
            return unavailable(version: version, reason: requestError.localizedDescription)
        }
        if statusCode < 200 || statusCode >= 300 {
            return unavailable(version: version, reason: "http \(statusCode)")
        }
        if !parsedJson {
            return unavailable(version: version, reason: "json missing")
        }
        guard let upToDate = parseUpToDate(body["upToDate"]) else {
            return unavailable(version: version, reason: "json missing")
        }
        let currentVersion = body["current_version"] as? String ?? ""
        let downloadUrl = body["download_url"] as? String ?? ""
        let available = (upToDate == false) && !downloadUrl.isEmpty
        return BridgeResponse.success(data: [
            "available": available,
            "upToDate": upToDate,
            "current_version": currentVersion,
            "download_url": downloadUrl,
            "version": currentVersion,
            "url": downloadUrl
        ])
    }
}
